File: 05 - SOLID Principles/Interface Segregation Principle/Java Code/FollowsISP.java
// =============================================================================
// INTERFACE SEGREGATION PRINCIPLE — the FIX (Java twin of follows-isp.cpp).
// Three small interfaces; a device implements ONLY what it really supports.
//
// Build & run:
//   javac FollowsISP.java
//   java FollowsISP
// =============================================================================
public class FollowsISP {

    interface Printer { void print(String doc); }
    interface Scanner { void scan(String doc); }
    interface Fax     { void fax(String doc); }

    // A basic printer implements ONLY Printer. It cannot lie about scanning.
    static class OfficePrinter implements Printer {
        public void print(String doc) { System.out.println("  [print] " + doc); }
    }

    // A high-end machine composes exactly the capabilities it really has.
    static class AllInOnePrinter implements Printer, Scanner, Fax {
        public void print(String doc) { System.out.println("  [print] " + doc); }
        public void scan(String doc)  { System.out.println("  [scan]  " + doc); }
        public void fax(String doc)   { System.out.println("  [fax]   " + doc); }
    }

    // Each client asks for exactly the interface it needs — nothing more.
    static void printDocument(Printer p, String doc) { p.print(doc); }
    static void archive(Scanner s, String doc) { s.scan(doc); }

    public static void main(String[] args) {
        OfficePrinter basic = new OfficePrinter();
        printDocument(basic, "invoice.pdf");
        // archive(basic, ...) won't compile — basic has no Scanner capability. Good.

        AllInOnePrinter combo = new AllInOnePrinter();
        printDocument(combo, "contract.pdf");
        archive(combo, "id-proof.jpg");
    }
}

File: 05 - SOLID Principles/Interface Segregation Principle/C++ Code/follows-isp.cpp
// =============================================================================
// INTERFACE SEGREGATION PRINCIPLE — the FIX.
// Three small, focused interfaces — Printer, Scanner, Fax. A device implements
// ONLY what it actually supports, so it can never be forced to fake a method.
//
// Build & run:
//   g++ -std=c++14 follows-isp.cpp -o follows
//   ./follows
// =============================================================================
#include <iostream>
#include <string>

using namespace std;

class Printer {
public:
    virtual void print(string doc) = 0;
    virtual ~Printer() {}
};

class Scanner {
public:
    virtual void scan(string doc) = 0;
    virtual ~Scanner() {}
};

class Fax {
public:
    virtual void fax(string doc) = 0;
    virtual ~Fax() {}
};

// A basic printer implements ONLY Printer. It cannot lie about scanning.
class OfficePrinter : public Printer {
public:
    void print(string doc) { cout << "  [print] " << doc << endl; }
};

// A high-end machine composes exactly the capabilities it really has.
class AllInOnePrinter : public Printer, public Scanner, public Fax {
public:
    void print(string doc) { cout << "  [print] " << doc << endl; }
    void scan(string doc) { cout << "  [scan]  " << doc << endl; }
    void fax(string doc) { cout << "  [fax]   " << doc << endl; }
};

// Each client asks for exactly the interface it needs — nothing more.
void printDocument(Printer* p, string doc) { p->print(doc); }
void archive(Scanner* s, string doc) { s->scan(doc); }

int main() {
    OfficePrinter* basic = new OfficePrinter();
    printDocument(basic, "invoice.pdf");
    // archive(basic, ...) won't compile — basic has no Scanner capability. Good.

    AllInOnePrinter* combo = new AllInOnePrinter();
    printDocument(combo, "contract.pdf");
    archive(combo, "id-proof.jpg");

    delete basic;
    delete combo;
    return 0;
}
